### PR TITLE
codegen: Do the same workaround we do for template parameters with `typename` on aliases for `decltypes` we can't resolve.

### PR DESCRIPTION
**Pull request by emilio**
r? @fitzgen

### CODE REVIEW

**Comment by emilio** [comment]
Actually, fitzgen was on vacation IIRC, want to review this @nox?

File: libbindgen/src/codegen/mod.rs [hunk 1]
```diff
@@ -516,6 +516,30 @@ impl CodeGenerator for Type {
...
+                    // with invalid template parameters, and at least this way
+                    // they can be replaced, instead of generating plain invalid
+                    // code.
+                    let inner_canon_type =
+                        inner_item.expect_type().canonical_type(ctx);
+                    if inner_canon_type.is_named() &&
+                        inner_canon_type.name().map_or(false, |name| {
+                            name.contains("decltype ") ||
+                            name.contains("decltype(")
+                        }) {
```
Comment (fitzgen) [inline libbindgen/src/codegen/mod.rs]: We should never have any spaces or parens for named types, regardless of decltype or typename.

What we want here is a generic check that this named type is a valid C++ identifier.

Comment (emilio) [inline libbindgen/src/codegen/mod.rs]: I tend to agree with this. My thinking was that given this is the exception, and not the norm, I think we want to know when can we generate these incorrect bits before skipping them.

I'm fine tweaking this to be a generic identifier check though, I guess.

**Comment by fitzgen** [comment]
@emilio were you going to circle back to this?

**Comment by emilio** [comment]
Whoops, forgot about this, there you go :)

File: src/ir/ty.rs [hunk 2]
```diff
@@ -292,6 +292,27 @@ impl Type {
...
+                let mut remaining = chars;
+
+                let valid = (first.is_alphabetic() || first == '_') &&
+                    remaining.all(|c| c.is_alphanumeric() || c == '_');
+
+                !valid
+            }
+            _ => false,
+        }
+    }
```
Comment (fitzgen) [inline src/ir/ty.rs]: Would be great to have a small `#[test]` unit test, or a doc test. Just for sanity.

Ok as a follow up, or as an issue marked E-Easy.

Comment (fitzgen) [inline src/ir/ty.rs]: r=me with your decision on what to do about ^

**Comment by emilio** [comment]
@bors-servo r=fitzgen

Will open an `E-easy` :)

**Pull request merged**